OpenRA.Mods.Common/Traits/Detonates.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Traits;
+
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class DetonatesInfo : ITraitInfo, Requires<RenderUnitInfo>
+	{
+		[Desc("Measured in ticks.")]  
+		public readonly int ChargeDelay = 0;
+		public readonly string ChargeSequence = null;
+		public readonly string ChargeSound = null;
+
+		[Desc("Measured in ticks.")]
+		public readonly int DetonationDelay = 0;
+		[WeaponReference]
+		public readonly string DetonationWeapon = null;
+
+		public object Create(ActorInitializer init) { return new Detonates(init.Self, this); }
+	}
+
+	public class Detonates : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly Actor self;
+		readonly DetonatesInfo info;
+		readonly RenderUnit renderUnit;
+		
+		public bool Deployed;
+
+		public Detonates(Actor self, DetonatesInfo info)
+		{
+			this.self = self;
+			this.info = info;
+			renderUnit = self.Trait<RenderUnit>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				yield return new TargetTypeOrderTargeter(new[] { "DetonateAttack" }, "DetonateAttack", 5, "attack", true, false) { ForceAttack = false };
+				yield return new DeployOrderTargeter("Detonate", 5);
+			}
+		}
+
+		void Detonate()
+		{
+			self.World.AddFrameEndTask(w =>
+			{
+				if (info.DetonationWeapon != null)
+				{
+					var weapon = self.World.Map.Rules.Weapons[info.DetonationWeapon.ToLowerInvariant()];
+
+					// Use .FromPos since this actor is killed. Cannot use Target.FromActor
+					weapon.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
+				}
+
+				self.Kill(self);
+			});
+		}
+
+		protected void StartDetonationSequence()
+		{
+			if (Deployed)
+				return;
+
+			Deployed = true;
+
+			if (info.ChargeSequence != null)
+				renderUnit.PlayCustomAnimationRepeating(self, info.ChargeSequence);
```

Comment:
Sorry, WithChargeOverlay. I remember @Mailaender had a pr for that for the TS obelisk.